doc/lightning-checkmessage.7
.\" -*- mode: troff; coding: utf-8 -*-
.TH "LIGHTNING-CHECKMESSAGE" "7" "" "Core Lightning 0.12.1" ""
.SH
NAME
.LP
lightning-checkmessage -- Command to check if a signature is from a node
.SH
SYNOPSIS
.LP
\fBcheckmessage\fR \fImessage\fR \fIzbase\fR [\fIpubkey\fR]
.SH
DESCRIPTION
.LP
The \fBcheckmessage\fR RPC command is the counterpart to
\fBsignmessage\fR: given a node id (\fIpubkey\fR), signature (\fIzbase\fR) and a
\fImessage\fR, it verifies that the signature was generated by that node
for that message (more technically: by someone who knows that node's
secret).
.PP
As a special case, if \fIpubkey\fR is not specified, we will try every
known node key (as per \fIlistnodes\fR), and verification succeeds if it
matches for any one of them.  Note: this is implemented far more
efficiently than trying each one, so performance is not a concern.
.PP
On failure, an error is returned and core lightning exit with the following error code:
- -32602: Parameter missed or malformed;
- 1301: \fIpubkey\fR not found in the graph.
.SH
RETURN VALUE
.LP
On success, an object is returned, containing:
.IP "\(bu" 2
\fBverified\fR (boolean): whether the signature was valid (always \fItrue\fR)
.if n \
.sp -1
.if t \
.sp -0.25v
.IP "\(bu" 2
\fBpubkey\fR (pubkey): the \fIpubkey\fR parameter, or the pubkey found by looking for known nodes
.SH
AUTHOR
.LP
Rusty Russell <\fIrusty@rustcorp.com.au\fR> is mainly responsible.
.SH
SEE ALSO
.LP
lightning-signmessage(7)
.SH
RESOURCES
.LP
Main web site: \fIhttps://github.com/ElementsProject/lightning\fR
\" SHA256STAMP:0d5c295293c006102476b93c23f8d3c870520419a8257118938e065161403148


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--
TODO: Insert version codename, and username of the contributor that named the release.
-->

## [0.12.1] - 2022-09-13: Web-8 init (dot one)

Point release with some bugfixes and patches.

### Added

None

### Changed

None

### Deprecated

None

### Removed

- build: `mrkd` and `mistune` not required to build project

### Fixed

- lnprototest: builds for lnprototest tests now use 22.04 LTS, which fixes a problem with loading `mako`. ([#5583])
- Plugins: topology plugin could crash when it sees duplicate private channel announcements ([#5593])
- connectd: proper `gossip_store` operation may resolve some previous gossip propagation issues and connectd crashes ([#5591])

[#5583]: https://github.com/ElementsProject/lightning/pull/5583
[#5593]: https://github.com/ElementsProject/lightning/pull/5593
[#5591]: https://github.com/ElementsProject/lightning/pull/5591


## [0.12.0] - 2022-08-23: Web-8 init

This release named by @adi2011.

Developers please note the Great Msat Migration has begun:
1. All JSON amount field names now end in "_msat" (others are deprecated)
2. Their values are strings ending in "msat", but will soon be normal integers.
3. You should accept both: set `allow-deprecated-apis=false` to test!

### Added

 - *NEW*: `commando` a new builtin plugin to send/recv peer commands over the lightning network, using runes. ([#5370])
 - *NEW*: New built-in plugin `bookkeeper` w/ commands `bkpr-listaccountevents`, `bkpr-listbalances`, `bkpr-listincome`, `bkpr-channelsapy`, `bkpr-dumpincomecsv`, `bkpr-inspect` ([#5071])
 - *NEW*: Emergency channel backup ("static backup") which allows us to seek fund recovery from honest peers in case of complete data loss ([#5422])
 - Config: `log-level=debug:<partial-nodeid>` supported to get debug-level logs for everything about a peer. ([#5349])
 - JSON-RPC: `connect` use the standard port derivation when the port is not specified. ([#5242])
 - JSON-RPC: `fetchinvoice` `changes` `amount_msat` ([#5306])
 - JSON-RPC: Added `mindepth` argument to specify the number of confirmations we require for `fundchannel` and `multifundchannel` ([#5275])
 - JSON-RPC: `listpeers` new fields for `funding` (`remote_funds_msat`, `local_funds_msat`, `fee_paid_msat`, `fee_rcvd_msat`). ([#5477])
 - JSON-RPC: `listpeers` add optional `remote_addr` ([#5244])
 - JSON-RPC: `listforwards` now shows `out_channel` in more cases: even if it couldn't actually send to it. ([#5330])
 - JSON-RPC: `pay` `attempts` `amount_msat` field. ([#5306])
 - Protocol: private channels will only route using short-channel-ids if channel opened with option_scid_alias-supporting peer. ([#5501])
 - Protocol: invoice routehints will use fake short-channel-ids for private channels if channel opened with option_scid_alias-supporting peer. ([#5501])
 - Protocol: we now advertize the `option_channel_type` feature (which we actually supported since v0.10.2) ([#5455])
 - Plugins: `channel_state_changed` now triggers for a v1 channel's initial "CHANNELD_AWAITING_LOCKIN" state transition (from prior state "unknown") ([#5381])
 - Plugins: `htlc_accepted_hook` `amount_msat` field. ([#5306])
 - Plugins: `htlc_accepted` now exposes the `short_channel_id` for the channel from which that HTLC is coming from and the low-level per-channel HTLC `id`, which are necessary for bridging two different Lightning Networks when MPP is involved. ([#5303])
 - Plugins: The `openchannel` hook may return a `mindepth` indicating how many confirmations are required. ([#5275])
 - msggen: introduce chain of responsibility pattern to make msggen extensible ([#5216])
 - cln_plugin: persist cln configuration from init msg ([#5279])
 - pyln-testing: Added utilities to read and parse `gossip_store` file for nodes. ([#5275])
 - `hsmtool`: new command `checkhsm` to check BIP39 passphrase against hsm_secret. ([#5441])
 - contrib: Added `fund_ln` to the contrib/startup\_regtest.sh ([#5062])
 - build: Added m1 architecture support for macos ([#4988])
 - build: Reproducible builds now include rust binaries such as the `cln-grpc` plugin ([#5421])


### Changed

 - `lightningd`: will refuse to start with the wrong node_id (i.e. hsm_secret changes). ([#5425])
 - `connectd`: prefer IPv6 connections when available. ([#5244])
 - `connectd`: Only use IP discovery as fallback when no addresses would be announced ([#5344])
 - `connectd`: give busy peers more time to respond to pings. ([#5347])
 - `gossipd`: now accepts spam gossip, but squelches it for ([#5239])
 - gossip: gossip\_store updated to version 10. ([#5239])
 - Options: `log-file` option specified multiple times opens multiple log files. ([#5281])
 - JSON-RPC: `sendpay` and `sendonion` now obey the first hop "channel" short_channel_id, if specified. ([#5505])
 - JSON-RPC: `signpsbt` no longer crashes if it doesn't like what your PSBT is ([#5506])
 - JSON-RPC: `signpsbt` will now add redeemscript + witness-utxo to the PSBT for an input that we can sign for, before signing it. ([#5506])
 - JSON-RPC: `plugin start` now assumes relative path to default plugins dir if the path is not found in absolute context. i.e. lightning-cli plugin start my_plugin.py ([#5211])
 - JSON-RPC: `fundchannel`: now errors if you try to buy a liquidity ad but dont' have `experimental-dual-fund` enabled ([#5389])
 - JSON-RPC: "\_msat" fields can be raw numbers, not "123msat" strings (please handle both!) ([#5306])
 - JSON-RPC: `invoice`, `sendonion`, `sendpay`, `pay`, `keysend`, `fetchinvoice`, `sendinvoice`: `msatoshi` argument is now called `amount_msat` to match other fields. ([#5306])


### Deprecated

Note: You should always set `allow-deprecated-apis=false` to test for changes.

 - JSON-RPC: `listpeers`.`funded` fields `local_msat` and `remote_msat`. ([#5477])

devtools/lightning-checkmessage.c
#include "config.h"
#include <bitcoin/pubkey.h>
#include <bitcoin/shadouble.h>
#include <ccan/err/err.h>
#include <common/bech32.h>
#include <common/utils.h>
#include <secp256k1_recovery.h>
#include <stdio.h>
#include <wally_core.h>

/* These tables copied from zbase32 src:
 * copyright 2002-2007 Zooko "Zooko" Wilcox-O'Hearn
 * mailto:zooko@zooko.com
 *
 * Permission is hereby granted to any person obtaining a copy of this work to
 * deal in this work without restriction (including the rights to use, modify,
 * distribute, sublicense, and/or sell copies).
 */

/* revchars: index into this table with the ASCII value of the char.  The result is the value of that quintet. */
static const u8 zbase32_revchars[]={ 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 18, 255, 25, 26, 27, 30, 29, 7, 31, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 24, 1, 12, 3, 8, 5, 6, 28, 21, 9, 10, 255, 11, 2, 16, 13, 14, 4, 22, 17, 19, 255, 20, 15, 0, 23, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, };

static void usage(void)
{
	fprintf(stderr, "Usage: lightning-checkmessage message zbase32-sig [key]\n"
		"If key does not match, signature is not valid!\n");
	exit(1);
}

int main(int argc, char *argv[])
{
	u8 *u5sig, *u8sig;
	size_t len;
	secp256k1_ecdsa_recoverable_signature rsig;
	struct sha256_ctx sctx = SHA256_INIT;
	struct sha256_double shad;
	struct pubkey reckey;
	const char *keystr;

	setup_locale();
	err_set_progname(argv[0]);
	wally_init(0);
	secp256k1_ctx = wally_get_secp_context();

	if (argc != 3 && argc != 4)
		usage();

	u5sig = tal_arr(NULL, u8, strlen(argv[2]));
	for (size_t i = 0; argv[2][i]; i++) {
		u5sig[i] = zbase32_revchars[(unsigned char)argv[2][i]];
		if (u5sig[i] > 31)
			errx(1, "Not a valid zbase32 string");
	}
	u8sig = tal_arr(NULL, u8, (strlen(argv[2]) * 5 + 7) / 8 + 1);
	len = 0;
	if (!bech32_convert_bits(u8sig, &len, 8, u5sig, strlen(argv[2]), 5, false))
		errx(1, "Invalid string");

	if (len != 65)
		errx(1, "Signature too %s", len < 65 ? "short" : "long");

	if (!secp256k1_ecdsa_recoverable_signature_parse_compact(secp256k1_ctx,
								 &rsig,
								 u8sig + 1,
								 u8sig[0] - 31))
		errx(1, "Signature not parsable");

	sha256_update(&sctx, "Lightning Signed Message:",
		      strlen("Lightning Signed Message:"));
	sha256_update(&sctx, argv[1], strlen(argv[1]));
	sha256_double_done(&sctx, &shad);

	if (!secp256k1_ecdsa_recover(secp256k1_ctx, &reckey.pubkey, &rsig,
				     shad.sha.u.u8))
		errx(1, "Signature not recoverable");

	keystr = pubkey_to_hexstr(NULL, &reckey);
	if (argv[3]) {
		if (!streq(keystr, argv[3]))
			errx(1, "Signature is invalid");
		printf("Signature is valid!\n");
	} else
		printf("Signature claims to be from key %s\n", keystr);
	return 0;
}
